akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala

```diff
@@ -0,0 +1,162 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.impl.fusing
+
+import akka.stream._
+import akka.stream.stage._
+import akka.stream.scaladsl._
+import akka.stream.actor.ActorSubscriberMessage
+import akka.stream.actor.ActorSubscriberMessage._
+import akka.stream.actor.ActorPublisherMessage
+import akka.stream.actor.ActorPublisherMessage._
+import scala.concurrent.forkjoin.ThreadLocalRandom
+
+import java.{ util ⇒ ju }
+
+class FlattenMerge[T, M](breadth: Int) extends GraphStage[FlowShape[Source[T, M], T]] {
+  private val in = Inlet[Source[T, M]]("flatten.in")
+  private val out = Outlet[T]("flatten.out")
+  override val shape = FlowShape(in, out)
+
+  override def createLogic(attr: Attributes) = new GraphStageLogic(shape) {
+
+    var sources = Set.empty[LocalSource[T]]
+    def activeSources = sources.size
+
+    private trait Queue {
+      def hasData: Boolean
+      def enqueue(src: LocalSource[T]): Unit
+      def dequeue(): LocalSource[T]
+    }
+
+    private class FixedQueue extends Queue {
+      final val Size = 16
+      final val Mask = 15
+
+      private val queue = new Array[LocalSource[T]](Size)
+      private var head = 0
+      private var tail = 0
+
+      def hasData = tail != head
+      def enqueue(src: LocalSource[T]): Unit =
+        if (tail - head == Size) {
+          val queue = new ju.LinkedList[LocalSource[T]]
+          while (hasData) {
+            queue.add(dequeue())
+          }
+          queue.add(src)
+          q = new DynamicQueue(queue)
+        } else {
+          queue(tail & Mask) = src
+          tail += 1
+        }
+      def dequeue(): LocalSource[T] = {
+        val ret = queue(head & Mask)
+        head += 1
+        ret
+      }
+    }
+
+    private class DynamicQueue(queue: ju.Queue[LocalSource[T]]) extends Queue {
```

Comment:
May as well extend ju.LinkedList to avoid the indirection (and It's only used in one place anyway)